compiler/cli/src/org/jetbrains/kotlin/cli/jvm/plugins/DependencyShadingPluginClassLoader.kt

```diff
@@ -0,0 +1,66 @@
+package org.jetbrains.kotlin.cli.jvm.plugins
+
+import org.jetbrains.org.objectweb.asm.ClassReader
+import org.jetbrains.org.objectweb.asm.ClassWriter
+import org.jetbrains.org.objectweb.asm.commons.ClassRemapper
+import org.jetbrains.org.objectweb.asm.commons.Remapper
+import java.io.IOException
+import java.io.InputStream
+import java.net.URL
+import java.net.URLClassLoader
+import java.security.AccessControlContext
+import java.security.AccessController
+import java.security.PrivilegedActionException
+import java.security.PrivilegedExceptionAction
+
+/**
+ * DependencyShadingPluginClassLoader will load a plugin jar, applying the necessary shading as required
+ * by the variant of the Kotlin Compiler which is consuming the plugin.
+ */
+class DependencyShadingPluginClassLoader(urls: Array<URL>, parent: ClassLoader) : URLClassLoader(urls, parent) {
+
+    val accessControlContext: AccessControlContext
+    val bytecodeLoader = URLClassLoader(urls, null)
+
+    init {
+        accessControlContext = AccessController.getContext()
+    }
+
+    // True if the Kotlin Compiler loading this plugin is shaded, and thus the plugin jar should also be shaded
+    val shouldBeShaded = com.intellij.mock.MockProject::class.java.name.startsWith("org.jetbrains.kotlin.")
+
+    @Throws(ClassNotFoundException::class)
+    override fun findClass(name: String): Class<*>? {
+        return try {
+            AccessController.doPrivileged(
+                PrivilegedExceptionAction {
+                    val path = name.replace('.', '/') + ".class"
+                    val originalBytes = bytecodeLoader.getResourceAsStream(path).readBytes()
+
+                    try {
+                        val cw = ClassWriter(ClassWriter.COMPUTE_MAXS)
+
+                        val shader = ClassRemapper(cw, object : Remapper() {
+                            override fun map(internalName: String): String {
+                                if(shouldBeShaded && internalName.startsWith("com/intellij/"))
```

Comment:
As far as I know, `com.intellij` is the only prefix that is a part of the compiler's public API, thus the only one we need to get right in order for plugins to easily support both environments.

The shading of other libraries within the Kotlin compiler has no adverse effects. Plugins which want to depend upon their own implementations of these libraries can themselves shade those libraries under their own package name, which will ensure there are never any conflicts.

Reply:
I tend to agree, it is good to keep it simple. But I wonder whether the approach may break anything, e.g. some plugins in the kotlin repo that happen to depend on the other relocated packages. While we always use the appropriate (shaded/unshaded) version of them, it is fine, though.

Reply:
Yeah, hopefully everyone is already following best practices, but even if they aren't, those existing jars will continue to work (we don't make the situation worse for them).  Hopefully they will want to ship a single jar that works in both variants, and this will help nudge plugin vendors to follow best-practice of not treating internal compiler dependencies as plugin dependencies.  Getting them to follow best practices has other added bonuses, like the plugins not being susceptible to dependency version API skew and treeshaking and other issues that could trip them up, so this is a good direction to nudge them anyway.